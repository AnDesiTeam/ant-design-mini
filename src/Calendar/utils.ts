import dayjs, { Dayjs } from 'dayjs';
import isoWeek from 'dayjs/plugin/isoWeek';
import { CellState, LocaleText, SelectionMode, CalendarValue } from './props';

dayjs.extend(isoWeek);

export function getMonthListFromRange(start: Dayjs, end: Dayjs): Dayjs[] {
  if (start.isAfter(end))
    throw new Error(`Start time can't be later than end time.`);
  const result = [];
  let current = start.date(1);
  while (!current.isAfter(end)) {
    result.push(current.date(1));
    current = current.add(1, 'month');
  }
  return result;
}

export function defaultMonthRange(): [number, number] {
  const start = dayjs().startOf('date');
  const end = dayjs().startOf('date').add(5, 'month');
  return [start.toDate().getTime(), end.toDate().getTime()];
}

/**
 *
 * @param month 月份的某一天
 * @param weekStartsOn 日历以星期几开始
 * @returns 获取当月日历所有的日子
 */
export function getDate(month: Dayjs, weekStartsOn: string): Dayjs[] {
  const startOfMonth = month.date(1);
  const cells: Dayjs[] = [];
  let iterator: Dayjs = startOfMonth
    .subtract(startOfMonth.isoWeekday() % 7, 'day')
    .startOf('day');

  if (weekStartsOn === 'Monday') {
    iterator = iterator.add(1, 'day');
    if (
      iterator.isSame(startOfMonth, 'month') &&
      !iterator.isSame(startOfMonth.startOf('date'), 'date')
    ) {
      iterator = iterator.add(-7, 'days');
    }
  }
  const diffDay = startOfMonth.date(1).add(1, 'month').diff(iterator, 'day');
  const lintCount = Math.ceil(diffDay / 7);
  while (cells.length < lintCount * 7) {
    cells.push(iterator);
    iterator = iterator.add(1, 'day');
  }
  return cells;
}

export function renderCells(
  cellsMonth: Dayjs,
  weekStartsOn: string,
  value: CalendarValue,
  localeText: LocaleText
): CellState[] {
  let rowBeginDay = 0;
  let rowEndDay = 6;
  if (weekStartsOn === 'Monday') {
    rowBeginDay = 1;
    rowEndDay = 0;
  }
  const dates = getDate(cellsMonth, weekStartsOn);
  if (!value) {
    return dates.map((d): CellState => {
      const isToday = dayjs().isSame(d, 'day');
      const isRowBegin =
        d.isSame(cellsMonth.startOf('month'), 'date') ||
        d.day() === rowBeginDay;
      const isRowEnd =
        d.isSame(cellsMonth.endOf('month'), 'date') || d.day() === rowEndDay;

      let top: CellState['top'];
      if (isToday) {
        top = {
          label: localeText.today,
        };
      }

      return {
        disabled: false,
        time: d.toDate().getTime(),
        date: d.get('date'),
        isSelect: false,
        isBegin: false,
        top,
        isEnd: false,
        inThisMonth: d.month() === cellsMonth.month(),
        isRowBegin,
        isRowEnd,
      };
    });
  }

  let selectBegin: Dayjs;
  let selectEnd: Dayjs;

  if (Array.isArray(value)) {
    selectBegin = dayjs(value[0]);
    selectEnd = dayjs(value[1] ?? value[0]);
  } else {
    selectBegin = dayjs(value);
    selectEnd = dayjs(value);
  }

  return dates.map((d): CellState => {
    const isToday = dayjs().isSame(d, 'day');
    const isRowBegin =
      d.isSame(cellsMonth.startOf('month'), 'date') || d.day() === rowBeginDay;
    const isRowEnd =
      d.isSame(cellsMonth.endOf('month'), 'date') || d.day() === rowEndDay;

    const isBegin = selectBegin.isSame(d, 'day');
    const isEnd = selectEnd.isSame(d, 'day');
    const isSelect =
      (!!selectBegin.isBefore(d, 'day') && !!selectEnd.isAfter(d, 'day')) ||
      isBegin ||
      isEnd;
    const inThisMonth = d.month() === cellsMonth.month();
    const time = d.toDate().getTime();
    let topLabel = isToday ? localeText.today : '';
    if (Array.isArray(value)) {
      if (isBegin) {
        if (isEnd && value.length === 2) {
          topLabel = localeText.startAndEnd;
        } else {
          topLabel = localeText.start;
        }
      } else {
        if (isEnd) {
          topLabel = localeText.end;
        }
      }
    }

    return {
      disabled: false,
      time,
      date: d.get('date'),
      isSelect,
      isBegin,
      top: { label: topLabel },
      isEnd,
      inThisMonth,
      isRowBegin,
      isRowEnd,
    };
  });
}

export function getSelectionModeFromValue(
  value?: CalendarValue
): SelectionMode {
  if (Array.isArray(value)) {
    return 'range';
  }
  if (typeof value === 'number') {
    return 'single';
  }
  return null;
}
