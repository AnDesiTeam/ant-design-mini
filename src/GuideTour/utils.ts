const buttonGroupKeys = [
  'nextButtonVisible',
  'lastButtonVisible',
  'jumpButtonVisible',
  'konwButtonVisible',
];

export const changeButtonVisible = (v: string[]) => {
  const newData = {};
  buttonGroupKeys.forEach((key) => {
    if (v.includes(key)) {
      newData[key] = true;
    } else {
      newData[key] = false;
    }
  });
  return newData;
};
