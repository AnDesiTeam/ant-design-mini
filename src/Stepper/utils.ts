/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-explicit-any */
// '1.' '1x' 'xx' '' => are not complete numbers
export function isNotCompleteNumber(num: any): boolean {
  return (
    isNaN(num) ||
    num === '' ||
    num === null ||
    (num && num.toString().indexOf('.') === num.toString().length - 1)
  );
}

export function toNumber(num: string, precision?: number): undefined | number {
  // num.length > 16 => This is to prevent input of large numbers
  const numberIsTooLarge = num && num.length > 16;
  if (isNotCompleteNumber(num) || numberIsTooLarge) {
    return undefined;
  }

  if (precision != null) {
    return Math.round(+num * Math.pow(10, precision) / Math.pow(10, precision));
  }
  return Number(num);
}

export function getPrecision(value: any, precision?: number): number {
  if (precision != null) {
    return precision;
  }
  const valueString = String(value);
  if (valueString.indexOf('e-') >= 0) {
    return parseInt(valueString.slice(valueString.indexOf('e-') + 2), 10);
  }
  let p = 0;
  if (valueString.indexOf('.') >= 0) {
    p = valueString.length - valueString.indexOf('.') - 1;
  }
  return p;
}

// step={1.0} value={1.51}
// press +
// then value should be 2.51, rather than 2.5
// if this.props.precision is undefined
// https://github.com/react-component/input-number/issues/39
export function getMaxPrecision(
  currentValue: any,
  step?: number,
  precision?: number,
): number {
  if (precision != null) {
    return precision;
  }

  const stepPrecision = getPrecision(step, precision);
  const currentValuePrecision = getPrecision(currentValue, precision);
  if (!currentValue) {
    return stepPrecision;
  }
  return Math.max(currentValuePrecision, stepPrecision);
}

export function getPrecisionFactor(currentValue: any, precision?: number): number {
  const p = getMaxPrecision(currentValue, undefined, precision);
  return Math.pow(10, p);
}

export function upStep(val: any, step?: number, precision?: number): number {
  const precisionFactor = getPrecisionFactor(val, precision);
  const p = Math.abs(getMaxPrecision(val, step, precision));
  const result = (
    (precisionFactor * val + precisionFactor * (step as number)) /
    precisionFactor
  ).toFixed(p);
  return toNumber(result, precision);
}

export function downStep(val: any, step?: number, precision?: number): number {
  const precisionFactor = getPrecisionFactor(val, precision);
  const p = Math.abs(getMaxPrecision(val, step, precision));
  const result = (
    (precisionFactor * val - precisionFactor * (step as number)) /
    precisionFactor
  ).toFixed(p);
  return toNumber(result);
}
