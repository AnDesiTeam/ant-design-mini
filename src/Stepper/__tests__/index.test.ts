import { getInstance } from '../../../tests/utils';
import fmtEvent from '../../_util/fmtEvent';
import { downStep, getMaxPrecision, getPrecision, upStep } from '../utils';

const component2 = {
  canIUse: (name) => name === 'component2',
};

describe('stepper', () => {
  it('load stepper', async () => {
    const instance = getInstance('Stepper', {}, component2);
    const instanceData = instance.getData();
    expect(instanceData).toMatchSnapshot();
  });
  it('初始值0，步长 0.01', async () => {
    const instance = getInstance('Stepper', { defaultValue: 0, step: 0.01 }, component2);
    const instanceData = instance.getData();
    expect(instanceData).toMatchSnapshot();

    instance.callMethod('onTap', fmtEvent({
      'data-mode': 'add',
    }));
    expect(instance.getData()).toMatchSnapshot();

    instance.callMethod('onTap', fmtEvent({
      'data-mode': 'minus',
    }));
    expect(instance.getData()).toMatchSnapshot();


    instance.callMethod('onTap', fmtEvent({
      'data-mode': 'add',
    }));
    expect(instance.getData()).toMatchSnapshot();

    instance.callMethod('onTap', fmtEvent({
      'data-mode': 'minus',
    }));
    expect(instance.getData()).toMatchSnapshot();
  });

  it('限制范围 [0,5]', async () => {
    const instance = getInstance('Stepper', { defaultValue: 0, step: 1, min: 0, max: 5 }, component2);
    const instanceData = instance.getData();
    expect(instanceData).toMatchSnapshot();

    let i = 0;
    while (++i <= 5) {
      instance.callMethod('onTap', fmtEvent({ 'data-mode': 'add', }));
    }
    expect(instance.getData()).toMatchSnapshot();

    // 不会再增加
    instance.callMethod('onTap', fmtEvent({ 'data-mode': 'add', }));
    expect(instance.getData()).toMatchSnapshot();

    let j = 0;
    while (++j <= 5) {
      instance.callMethod('onTap', fmtEvent({ 'data-mode': 'minus', }));
    }
    expect(instance.getData()).toMatchSnapshot();

    // 不会再减少
    instance.callMethod('onTap', fmtEvent({ 'data-mode': 'minus', }));
    expect(instance.getData()).toMatchSnapshot();
  });

  it('受控组件', async () => {
    const onChange = jest.fn();
    const instance = getInstance('Stepper', { value: 0, onChange, }, component2);

    instance.callMethod('onTap', fmtEvent({ 'data-mode': 'add', }));
    expect(instance.getData()).toMatchSnapshot();

    expect(onChange).toBeCalledTimes(1);
    expect(onChange.mock.calls[0][0]).toBe(1);

    instance.callMethod('onTap', fmtEvent({ 'data-mode': 'minus', }));
    expect(instance.getData()).toMatchSnapshot();
    expect(onChange).toBeCalledTimes(2);
    expect(onChange.mock.calls[0][0]).toBe(1);

    instance.setProps({ value: 10 });
    expect(instance.getData()).toMatchSnapshot();
  });

  it('变更输入框', async () => {
    const instance = getInstance('Stepper', { value: 0 }, component2);
    expect(instance.getData()).toMatchSnapshot();

    const inputAndExpct = [
      ['1', '1'],
      ['1.1', '1.1'],
      ['1.1.4', '1.1'],
      ['1.15', '1.15'],
    ];

    for (let item of inputAndExpct) {
      instance.callMethod('onChange', item[0]);
      expect(instance.getData().mixin.value).toEqual(item[1]);
    }
  });

  it('计算精度的 utils 方法', async () => {
    expect(getPrecision(1)).toBe(0);
    expect(getPrecision(1.1)).toBe(1);
    expect(getPrecision(1.23)).toBe(2);

    expect(getMaxPrecision(1.23, 0.1)).toBe(2);
    expect(getMaxPrecision(1.23, 0.1, 1)).toBe(1);
    expect(getMaxPrecision(1.23, undefined, 1)).toBe(1);

    // expect(upStep(1.53, 0.1, 1)).toBe(1.6); // 没跑过？
    // expect(upStep(1.23, 0.1, 1)).toBe(1.3); // 没跑过？
  });

  it('计算精度', async () => {
    // const instance = getInstance('Stepper', { value: 1.23, precision: 2, step: 0.1 }, component2);
    // expect(instance.getData().mixin.value).toEqual('1.23'); // 感觉不对？
    // instance.callMethod('onTap', fmtEvent({ 'data-mode': 'add', }));
    // expect(instance.getData().mixin.value).toEqual('1.234');
  });
});
