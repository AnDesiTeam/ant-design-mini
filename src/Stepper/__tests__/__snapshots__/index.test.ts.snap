// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stepper load stepper 1`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "",
  },
}
`;

exports[`stepper 初始值0，步长 0.01 1`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 初始值0，步长 0.01 2`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0.01",
  },
}
`;

exports[`stepper 初始值0，步长 0.01 3`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 初始值0，步长 0.01 4`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0.01",
  },
}
`;

exports[`stepper 初始值0，步长 0.01 5`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 受控组件 1`] = `
Object {
  "mixin": Object {
    "controlled": true,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 受控组件 2`] = `
Object {
  "mixin": Object {
    "controlled": true,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 受控组件 3`] = `
Object {
  "mixin": Object {
    "controlled": true,
    "updated": true,
    "value": "10",
  },
}
`;

exports[`stepper 变更输入框 1`] = `
Object {
  "mixin": Object {
    "controlled": true,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 限制范围 [0,5] 1`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 限制范围 [0,5] 2`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "5",
  },
}
`;

exports[`stepper 限制范围 [0,5] 3`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "5",
  },
}
`;

exports[`stepper 限制范围 [0,5] 4`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;

exports[`stepper 限制范围 [0,5] 5`] = `
Object {
  "mixin": Object {
    "controlled": false,
    "updated": true,
    "value": "0",
  },
}
`;
